doc/sphinx/proof-engine/ltac2.rst

```diff
@@ -659,168 +734,476 @@ insert in a concise way an Ltac2 variable of type :n:`constr` into a Coq term.
 Match over terms
 ~~~~~~~~~~~~~~~~
 
-Ltac2 features a construction similar to Ltac1 :n:`match` over terms, although
+Ltac2 features a construction similar to Ltac1 :tacn:`match` over terms, although
 in a less hard-wired way.
 
-.. productionlist:: coq
-   ltac2_term       : match! `ltac2_term` with `constrmatching` .. `constrmatching` end
-                    : lazy_match! `ltac2_term` with `constrmatching` .. `constrmatching` end
-                    : multi_match! `ltac2_term` with `constrmatching` .. `constrmatching` end
-   constrmatching  : | `constrpattern` => `ltac2_term`
-   constrpattern   : `term`
-                   : context  [ `term` ]
-                   : context `lident` [ `term` ]
-
-This construction is not primitive and is desugared at parsing time into
-calls to term matching functions from the `Pattern` module. Internally, it is
-implemented thanks to a specific scope accepting the :n:`@constrmatching` syntax.
-
-Variables from the :n:`@constrpattern` are statically bound in the body of the branch, to
+.. tacn:: @ltac2_match_key @ltac2_expr__term with @ltac2_match_list end
+   :name: lazy_match!; match!; multi_match!
+
+   .. insertprodn ltac2_match_key ltac2_match_pattern
+
+   .. prodn::
+      ltac2_match_key ::= lazy_match!
+      | match!
+      | multi_match!
+      ltac2_match_list ::= {? %| } {+| @ltac2_match_rule }
+      ltac2_match_rule ::= @ltac2_match_pattern => @ltac2_expr
+      ltac2_match_pattern ::= @cpattern
+      | context {? @ident } [ @cpattern ]
+
+   Evaluates :n:`@ltac2_expr__term`, which must yield a term, and matches it
+   sequentially with the :token:`ltac2_match_pattern`\s, which may have
+   metavariables.  When a match is found, metavariable values are substituted
+   into :n:`@ltac2_expr`, which is then applied.
+
+   Matching may continue depending on whether  `lazy_match!`, `match!` or `multi_match!`
+   is specified.
+
+   .. todo PR I see ? is sometimes need in the ltac2_expr and sometimes not, but
+      I don't when which is required
+
+   NOT RIGHT YET: In the :token:`ltac2_match_pattern`\s, metavariables have the form :n:`?@ident`, whereas
+   in the :n:`@ltac2_expr`\s, the question mark is omitted.  Choose your metavariable
+   names with care to avoid name conflicts.  For example, if you use the metavariable `S`,
+   then the :token:`ltac2_expr` can't use `S` to refer to the constructor of `nat`
+   without qualifying the constructor as `Datatypes.S`.
+
+   .. todo below: is matching non-linear unification?  is it the same or different
+      from unification elsewhere in Coq?
```

Comment:
This is a question for @ppedrot ; I have not tested the non-linearity of Ltac2 against Ltac1

Reply:
I raised the question for Ltac1 but didn't get a useful answer.

Reply:
It's the same underlying algorithm as Ltac1, and IIRC I've explained it on another issue, but for the sake of completeness:
- inside the same term, e.g. `match goal with [ |- ?P -> ?P ] => ... end` or `match goal with [ H : ?P -> ?P |- _ ] => ... end`  non-linear patterns are compared up to syntactic equality.
- across hypotheses, e.g. e.g. `match goal with [ H : ?P |- ?P ] => non-linear pattern are compared up to convertibility.

Reply:
@ppedrot Thanks.  I originally asked the question while working on the Ltac1 chapter.  Indeed, the text here is a cut-and-paste from that chapter.

> non-linear patterns are compared up to syntactic equality.

I don't understand "up to ...".  Do you mean it's a strictly syntactic comparison of the ASTs, i.e. "3" doesn't match "1+2" or is it more powerful than that?  And I expect the "non-linear" is not restrictive i.e. linear and non-linear patterns are processed the same way.  Right?

> non-linear pattern are compared up to convertibility.

Most unfortunately, I don't see a definition of "convertibility" anywhere in the doc.  It seems to be an essential concept, though, and if so, I will look the other way while I drop the term into the text, with the expectation we'll add a definition in the near future.

Your words mean: if (A is convertible into B or vice-versa **OR** A and B both convert to the same canonical form) then it's a match?

Reply:
> it's a strictly syntactic comparison

Syntactic equality includes α-conversion though, so e.g. it'll match `fun x : nat => x` with `fun y : nat => y`

> I expect the "non-linear" is not restrictive

It doesn't make sense to ask for the comparison algorithm for a linear pattern, since it'll never be called. But I guess it's a matter of phrasing.

>  I don't see a definition of "convertibility" anywhere in the doc. It seems to be an essential concept

It definitely is, and even has [its own section](https://coq.inria.fr/refman/language/core/conversion.html) in the documentation of typing rules.

> Your words mean

Conversion is an equivalence relation on well-typed terms, so you don't need the OR, just that "A is convertible to B".

Reply:
I don't see a definition of α-conversion anywhere in the documentation.  Seems like that should be in the conversion chapter, right?  I suppose we include α-conversion for almost every comparison of terms.  Should α be included in phrases such as: "We say that two terms t1 and t2 are βδιζη-convertible, or simply convertible, or equivalent"?

> It doesn't make sense to ask for the comparison algorithm for a linear pattern, since it'll never be called. But I guess it's a matter of phrasing.

Yeah, basic writing: don't introduce unnecessary terms.  If I say "non-linear pattern" that immediately suggests the question of how linear patterns are handled.  Of course that's the trivial case.  Not worth an extra sentence to cover that detail and would be a bit harder to understand.

Not sure how I missed the convertibility section.  I did a full-text search of the source and didn't find it.  Pilot error or an IDE glitch.

> Conversion is an equivalence relation on well-typed terms, so you don't need the OR, just that "A is convertible to B".

Is that the most common phrase?  From the description (below) "A and B are convertible" sounds more accurate to me.

> We say that two terms t1 and t2 are βδιζη-convertible, or simply convertible, or equivalent, in the global environment E and local context Γ iff there exist terms u1 and u2 such that E[Γ]⊢t1▹…▹u1 and E[Γ]⊢t2▹…▹u2 and either u1 and u2 are identical up to irrelevant subterms, or they are convertible up to η-expansion, i.e. u1 is λx:T. u′1 and u2x is recursively convertible to u′1, or, symmetrically, u2 is λx:T. u′2 and u1x is recursively convertible to u′2. We then write E[Γ]⊢t1=βδιζηt2.

Reply:
Also, I happened to notice this in https://coq.inria.fr/distrib/V8.12+beta1/stdlib/Ltac2.Constr.html

```
Ltac2 @ external equal : constr -> constr -> bool := "ltac2" "constr_equal".
Strict syntactic equality: only up to α-conversion and evar expansion
```

Is evar expansion perhaps also used for for unification or not?  If so, does it correspond to one of `βδιζη`?